### ISSUE TITLE
TypeError: undefined is not an object (evaluating 'ha_elements().tabs')

### ISSUE THREAD

**Issue by Mariusthvdb** [maykar/custom-header#159]
these errors are frequent in the HA-log:

2020-04-02 14:25:54 ERROR (MainThread) [frontend.js.latest.202003181] https://redacted.duckdns.org:port/local/lovelace/resources/custom-header.js?v=1.3.8:603:55085 TypeError: undefined is not an object (evaluating 'ha_elements().tabs')

in Inspector:

<img width="856" alt="Schermafbeelding 2020-04-02 om 14 34 06" src="https://user-images.githubusercontent.com/33354141/78249695-1239b900-74ef-11ea-85b5-66fb2ca62f8b.png">

Any suggestion to what I should be looking for in the tabs config?

thanks!
CH 1.3.8
HA 0.107.5

tabs related config:

```
show_tabs: >-
  {% if is_state('input_select.mode','Developer') %}
  {%- if 'Chrome' in userAgent -%}0 to 30
  {%- else -%}0 to 29
  {%- endif -%}
  {%- else -%}0 to 3,5 to 22,27,29
  {%- endif -%}
```

```
reverse_tab_direction: >-
  {{is_state('input_boolean.ch_reverse_tab_direction','on')}}

hidden_tab_redirect: >-
  {{is_state('input_boolean.ch_hidden_tab_redirect','on')}}
```
and
```
tab_icons:
  alarm: >-
    {{states('sensor.alarm_panel_icon')}}
  phones_tablets: >-
    {{states('sensor.presence_icon')}}
  weer_klimaat: >-
    {{states('sensor.weather_icon')}}
  alarmclock: >-
    {{'mdi:alarm-off' if is_state('sensor.next_alarm','Not set') else 'mdi:alarm'}}

#active_tab_color: black
active_tab_css: 'background-color: rgba(0, 0, 0, 0.3);'
tab_indicator_color: black
tabs_color:
  alarm: >-
    {{states('sensor.alarm_panel_icon_color')}}
  phones_tablets: >-
    {{state_attr('sensor.family_home','icon_color')}}
  alarmclock: >-
    {{'green' if is_state('sensor.next_alarm','Not set') else 'red'}}
  home_climate: >-
    {{states('sensor.temperature_color_name')}}
```

exceptions tab related:

```
  - conditions:
      user: Gezin, Names, name, name
    config:
      show_tabs: 0 to 3,7,16 to 19
      voice_hide: true
      options_hide: true
```

I've checked all templates and they evaluate as expected:

<img width="392" alt="Schermafbeelding 2020-04-02 om 14 37 36" src="https://user-images.githubusercontent.com/33354141/78250031-8e340100-74ef-11ea-9898-965c787d32cb.png">

**Comment by Mariusthvdb** [comment]
update, now see this in the HA log viewer, hope this give some more clues where to go looking:
<img width="1164" alt="Schermafbeelding 2020-04-03 om 14 11 07" src="https://user-images.githubusercontent.com/33354141/78359288-16c8a500-75b5-11ea-98ee-5316f8c6315b.png">

**Comment by Mariusthvdb** [comment]
is there any acknowledgement of this issue? I am still seeing in on 108.8 now, and no idea what to test other than completely disabling CH, which of course would be sub-optimal....

Ive checked and rechecked all individual templates but keep seeing:

<img width="556" alt="Schermafbeelding 2020-04-22 om 17 20 35" src="https://user-images.githubusercontent.com/33354141/80000578-99aa9480-84bd-11ea-803b-9613e509ba96.png">

using CH 1.3.8

homeassistent.log:
```
2020-04-22 17:08:18 ERROR (MainThread) [frontend.js.latest.202004072] https://redacted/local/lovelace/resources/custom-header.js?v=1.3.8:603:55085 TypeError: undefined is not an object (evaluating 'ha_elements().tabs')
```
please let me know what to do.

**Comment by maykar** [comment]
I'm aware of all issues posted here and on the forums, but I have [very little time](https://community.home-assistant.io/t/custom-header/155399/695?u=mayker) at the moment, please bear with me.

**Comment by Mariusthvdb** [comment]
of course! please keep up with things and stay healthy.

Btw, I might have accidentally solved my own issue here.... 

this is a template for a marquee I use:
```
header_text: >-
  {% if is_state('input_boolean.ch_use_header_text_template','on') %}
  {% if is_state('input_boolean.ch_vertical_marquee_up','on') %}
  {% set keyframe = "@keyframes show {
  100% {margin-top:-144px;}
  83% {margin-top:-96px;}
  67% {margin-top:-96px;}
  50% {margin-top:-48px;}
  33% {margin-top:-48px;}
  17% {margin-top:0px;}
  0% {margin-top:0px;}}" %}
  {% else %}
  {% set keyframe = "@keyframes show {
  0% {margin-top:-144px;}
  17% {margin-top:-96px;}
  33% {margin-top:-96px;}
  50% {margin-top:-48px;}
  67% {margin-top:-48px;}
  83% {margin-top:0px;}
  100% {margin-top:0px;}}" %}
  {% endif %}
  <style>
  #flipper{color:#999;}
  #flip {height:24px;overflow:hidden;}
  #flip > div > div {color:var(--primary-text-color);height:24px;margin-bottom:24px;display:inline-block;text-overflow: clip;white-space: nowrap;}
  #flip div:first-child {animation: show 10s linear infinite;}
  {{keyframe}}
  #flipper p {position:fixed;color:#999; }
  </style>
  <div style='float: left'>Ha Rpi4: {{viewTitle}}: ;</div>      #<---- note the lost ';' there
  <div id=flip><div><div>
  Summary - {{states('sensor.dark_sky_summary')}}</div></div><div><div>
  Daily - {{states('sensor.dark_sky_daily_summary')}}</div></div><div><div>
  Hourly - {{states('sensor.dark_sky_hourly_summary')}}</div></div><div><div>
  Summary - {{states('sensor.dark_sky_summary')}}</div></div><div><div>
  {% else %}Ha Rpi4: {{viewTitle}}
  {% endif %}
```

I fear by deleting a '&nbsp' in that line, I left the ';' in, which seems to have caused the issue. After having reinstated the line to
```
 <div style='float: left'>Ha Rpi4: {{viewTitle}}:&nbsp;</div>
```

The error hasn't re-appeared. I still dont understand why the error is what it is, or how I could have debugged from there.

For reference, if you find time, could this be it?

**Comment by Mariusthvdb** [comment]
update:
shouldn't have said it of course, the error is re-appearing.. so sorry, would have hoped it to be solved. Please keep open...

**Comment by Mariusthvdb** [comment]
Hi Ryan, 
not sure yet, since my production system is on 109.6 still, and several of the latest updates caused issues, so I reverted to an earlier version.

Is the latest update compatible with 109x?

**Comment by maykar** [comment]
I'm not entirely sure if the latest is 100% compatible with 109 and I just [made a post](https://community.home-assistant.io/t/custom-header/155399/803) on how I plan to drop backwards compatibility in favor of both development and optimization.

Steven_Rollason also just [made a post](https://community.home-assistant.io/t/custom-header/155399/805) explaining a workaround on how to use icons in header_text again until I get around to messing with that "experimental and unsupported" feature :)

The original error itself is still a bit of mystery to me (I'm thinking it is/was a race condition that may have been solved), but when you do feel comfortable with updating please keep me updated as quite a bit has changed in the last week. Feel free to keep this open until then.

**Comment by Mariusthvdb** [comment]
check, will do!

**Comment by maykar** [comment]
And 1.4.5 is displayed in the dev tools console as well?

**Comment by Mariusthvdb** [comment]
sure, updating works fine. now on 1.4.7:
<img width="714" alt="Schermafbeelding 2020-05-28 om 09 45 02" src="https://user-images.githubusercontent.com/33354141/83113458-f9e9b300-a0c7-11ea-8740-89f1a1f1b32b.png">

on ha 110.2/3 this has become even more prominent again, sorry for the long snippet, but I post it because maybe the timing might give you a clue...

```
2020-05-28 09:23:21 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:23:21 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:23:21 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:23:35 ERROR (MainThread) [plexwebsocket] Websocket disconnected
2020-05-28 09:23:46 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:23:46 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:23:46 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:24:13 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:24:13 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:24:13 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:24:43 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:24:43 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:24:43 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:25:14 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:25:14 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:25:14 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:25:44 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:25:44 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:25:44 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:01 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:01 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:01 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:24 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:24 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:24 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:57 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:57 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:26:57 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:27:16 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:27:16 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
2020-05-28 09:27:16 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
```

from developer-tools/logs
```
Log Details (ERROR)
Logger: frontend.js.latest.202005195
Source: components/system_log/__init__.py:209 
First occurred: 9:23:21 AM (30 occurrences) 
Last logged: 9:27:16 AM

https://mydomain.duckdns:port/local/lovelace/resources/custom-header/custom-header.js?v=1.4.7:596:20117 Uncaught TypeError: Cannot read property 'tabs' of undefined
```

**Comment by maykar** [comment]
Give the latest pre-release a try (1.4.9) and let me know if anything changes.

**Comment by Mariusthvdb** [comment]
thanks, I just did and now see 1 of these:
```
2020-05-28 18:02:02 ERROR (MainThread) [frontend.js.latest.202005195] https://mydomain/local/lovelace/resources/custom-header/custom-header.js?v=1.4.9:596:62551 TypeError: undefined is not an object (evaluating 'B.partialPanelResolver')
```

**Comment by maykar** [comment]
Going to add a fix for that one in the next release and close this one as the original issue has been resolved. Thanks!

**Comment by Mariusthvdb** [comment]
Great!
could you please let us know what caused this? I mean, after hunting this for some time, and checking all config templates, I really like to understand ;-)

**Comment by maykar** [comment]
Don't know the cause, but the symptom is at some point on your system the tab element's just aren't found. Since this happens quickly and often I imagine it's either template related or possibly a conflict with another addon. This has been an issue for you for almost 2 months and no one else has mentioned a similar issue, making it difficult to connect any dots on why this is happening. 

A few of the changes I've made over that time were attempts to fix any cause of the issue, but what I ended up doing in the end was essentially only run that bit of code when they are found (something I still don't think is necessary, but in your particular case it works).
